torchvision/prototype/datasets/_builtin/sintel.py

```diff
@@ -0,0 +1,167 @@
+        w = int.from_bytes(file.read(4), "little")
+        h = int.from_bytes(file.read(4), "little")
+        data = file.read(2 * w * h * 4)
+        data = np.frombuffer(data, dtype=np.float32)
+        return data.reshape(h, w, 2).transpose(2, 0, 1)
+
+    def _flows_key(self, data: Tuple[str, Any]) -> Tuple[str, int]:
+        path = pathlib.Path(data[0])
+        category = path.parent.name
+        idx = int(FILE_NAME_PATTERN.match(path.name).group("idx"))  # type: ignore[union-attr]
+        return category, idx
+
+    def _images_key(self, data: Tuple[Tuple[str, Any], Tuple[str, Any]]) -> Tuple[str, int]:
+        return self._flows_key(data[0])
+
+    def _collate_and_decode_sample(
+        self,
+        data: Tuple[Tuple[str, io.IOBase], Any],
+        *,
+        decoder: Optional[Callable[[io.IOBase], torch.Tensor]],
+        config: DatasetConfig,
+    ) -> Dict[str, Any]:
+        if config.split == "train":
+            flo, images = data
+            img1, img2 = images
+            flow_arr = self._read_flo(flo[1])
+            del images, flo
+        else:
+            # When split is `test`
+            img1, img2 = data
+
+        del data
+
+        path1, buffer1 = img1
+        path2, buffer2 = img2
+
+        return dict(
+            image1=decoder(buffer1) if decoder else buffer1,
+            image2=decoder(buffer2) if decoder else buffer2,
+            flow=flow_arr if config.split == "train" else None,
+        )
+
+    def _make_datapipe(
+        self,
+        resource_dps: List[IterDataPipe],
+        *,
+        config: DatasetConfig,
+        decoder: Optional[Callable[[io.IOBase], torch.Tensor]],
+    ) -> IterDataPipe[Tuple[Dict[str, Any], Dict[str, Any]]]:
+        dp = resource_dps[0]
+        archive_dp = ZipArchiveReader(dp)
+
+        curr_split = Filter(archive_dp, partial(self._classify_train_test, config=config))
+
+        pass_images_dp, flo_dp = Demultiplexer(
+            curr_split,
+            2,
+            partial(self._classify_archive, config=config),
+            drop_none=True,
+            buffer_size=INFINITE_BUFFER_SIZE,
+        )
+        flo_dp = Shuffler(flo_dp, buffer_size=INFINITE_BUFFER_SIZE)
+
+        pass_images_dp: IterDataPipe[Tuple[str, Any], Tuple[stry, Any]] = IntCategoryGrouper(pass_images_dp)
+        if config.split == "train":
+            zipped_dp = IterKeyZipper(
+                flo_dp,
+                pass_images_dp,
+                key_fn=self._flows_key,
+                ref_key_fn=self._images_key,
+            )
+            return Mapper(zipped_dp, self._collate_and_decode_sample, fn_kwargs=dict(decoder=decoder, config=config))
+        else:
+            # When split is `test`, flo_dp will be empty and thus should not be zipped with pass_images_dp
+            return Mapper(
+                pass_images_dp, self._collate_and_decode_sample, fn_kwargs=dict(decoder=decoder, config=config)
+            )
```

Comment:
I would restructure this as follows: We can branch (`if config.split == "train"`) directly after the filtering. In the train branch, we are going to split and merge the datapipe in order to get image / flow pairs. In the test branch we don't need to split anything, but just filter out the images. Note that they also need to be shuffled.